src/dns.c

```diff
@@ -2191,163 +2226,244 @@ hsk_dns_name_parse(
+        // Ensure enough room for 2-byte pointer, this should only
+        // be an issue if we are pointing to a 1-byte label...!
+        if (2 + noff > HSK_DNS_MAX_NAME) {
           *len = off;
           return false;
         }
-        data[off] = size;
-      }
 
-      if (cmp) {
-        char *sub = (char *)&name[begin];
-        if (strcmp(sub, ".") != 0) {
-          size_t p = (size_t)hsk_map_get(&cmp->map, sub);
-          if (p == 0) {
-            size_t o = (size_t)(&data[off] - cmp->msg);
-            if (o < (2 << 13))
-              hsk_map_set(&cmp->map, sub, (void *)o);
-          } else {
-            if (ptr == -1) {
-              ptr = p;
-              pos = off;
-              i += strlen(s);
-              break;
-            }
-          }
+        if (data) {
+          // Encode and write pointer
+          p ^= 0xc000;
+          data[off] = (p >> 8) & 0xff;
+          data[off + 1] = p & 0xff;
         }
+
+        // Done: pointers are always last
+        off += 2;
+        *len = off;
+        return true;
       }
+    }
 
-      off += 1;
+    // Compression is off, or the label is new: write it raw
+    if (data)
+      memcpy(&data[off], &name[noff], label + 1);
 
-      if (data) {
-        int j;
-        for (j = begin; j < i; j++) {
-          char ch = name[j];
+    off += label + 1;
+    noff += label + 1;
+  }
+}
 
-          // 0xff -> NUL
-          if (ch == -1)
-            ch = '\0';
+bool
+hsk_dns_name_to_string(const uint8_t *name, char *namestr) {
+  // Offset of `name` array being read
+  int off = 0;
+  // Offset of output `namestr` string being written
+  int noff = 0;
 
-          // 0xfe -> .
-          if (ch == -2)
-            ch = '.';
+  for (;;) {
+    uint8_t label = name[off];
 
-          data[off++] = ch;
-        }
+    if (label == 0x00) {
+      if (noff == 0) {
+        // Only one label and it's root
+        namestr[noff++] = '.';
+        namestr[noff] = '\0';
       } else {
-        off += size;
+        namestr[noff] = '\0';
       }
-
-      begin = i + 1;
+      return true;
     }
-  }
 
-  if (i > HSK_DNS_MAX_NAME) {
-    *len = off;
-    return false;
-  }
+    for (; label > 0; label--) {
+      off++;
+      uint8_t c = name[off];
+
+      switch (c) {
+        // Escape special characters
+        case 0x2e /*.*/:
+        case 0x28 /*(*/:
+        case 0x29 /*)*/:
+        case 0x3b /*;*/:
+        case 0x20 /* */:
+        case 0x40 /*@*/:
+        case 0x22 /*"*/:
+        case 0x5c /*\\*/:
+          sprintf(&namestr[noff], "\\%c", (char)c);
+          noff += 2;
+          continue;
+        default:
+          // Write escaped, three-digit byte code
+          if (c < 0x20 || c > 0x7e) {
+            sprintf(&namestr[noff], "\\%03d", c);
+            noff += 4;
+            continue;
+          }
+      }
 
-  if (i == 0 || name[i - 1] != '.') {
-    *len = off;
-    return false;
-  }
+      // Boring, printable character
+      namestr[noff++] = c;
```

Comment:
I think moving this inside `default` - after escaped char checks would make this more readable.

Reply:
ok cool added in f5200ab1251950b897e8ef3f71f5a11737aeaa5a